sdk/src/main/java/com/kin/ecosystem/Kin.java

```diff
@@ -109,100 +198,102 @@ private static SignInData getJwtSignInData(@NonNull final String jwt) {
 			.jwt(jwt);
 	}
 
-	private synchronized static void init(@NonNull Context appContext, @NonNull SignInData signInData,
-		@NonNull KinEnvironment environment) throws ClientException, BlockchainException {
-		Configuration.setEnvironment(environment);
-		instance = getInstance();
-		appContext = appContext.getApplicationContext(); // use application context to avoid leaks.
-		DeviceUtils.init(appContext);
-		initBlockchain(appContext);
-		initAuthRepository(appContext, signInData);
-		initEventCommonData(appContext);
-		instance.eventLogger.send(KinSdkInitiated.create());
-		initAccountManager(appContext);
-		initOrderRepository(appContext);
-		initOfferRepository();
-		setAppID();
-	}
-
-	private static void initAccountManager(@NonNull final Context context) {
-		AccountManagerImpl
-			.init(AccountManagerLocal.getInstance(context), instance.eventLogger, AuthRepository.getInstance(),
-				BlockchainSourceImpl.getInstance());
-		if (!AccountManagerImpl.getInstance().isAccountCreated()) {
-			AccountManagerImpl.getInstance().start();
+	private static void internalLogin(@NonNull SignInData signInData, final KinCallback<Void> loginCallback) {
+		String publicAddress = null;
+		try {
+			checkInstanceNotNull();
+			try {
+				BlockchainSourceImpl.getInstance().createAccount();
+				publicAddress = getPublicAddress();
+			} catch (final BlockchainException exception) {
+				instance.executorsUtil.mainThread().execute(new Runnable() {
+					@Override
+					public void run() {
+						loginCallback.onFailure(exception);
+					}
+				});
+			}
+		} catch (final ClientException exception) {
+			instance.executorsUtil.mainThread().execute(new Runnable() {
+				@Override
+				public void run() {
+					loginCallback.onFailure(exception);
+				}
+			});
 		}
-	}
 
-	private static void initEventCommonData(@NonNull Context context) {
-		EventCommonDataUtil.setBaseData(context);
-	}
+		String deviceID = AuthRepository.getInstance().getDeviceID();
+		signInData.setDeviceId(deviceID != null ? deviceID : UUID.randomUUID().toString());
+		signInData.setWalletAddress(publicAddress);
+		AuthRepository.getInstance().setSignInData(signInData);
 
-	private static void setAppID() {
 		ObservableData<String> observableData = AuthRepository.getInstance().getAppID();
 		String appID = observableData.getValue();
-		observableData.addObserver(new Observer<String>() {
-			@Override
-			public void onChanged(String appID) {
-				BlockchainSourceImpl.getInstance().setAppID(appID);
-			}
-		});
-
+		if (appID == null) {
+			observableData.addObserver(new Observer<String>() {
```

Comment:
do we need it to observer forever? or can we unsubscribe after the first event?

Reply:
👍